vm/systemSmartContracts/liquidStaking.go

```diff
@@ -0,0 +1,244 @@
+//go:generate protoc -I=proto -I=$GOPATH/src -I=$GOPATH/src/github.com/ElrondNetwork/protobuf/protobuf  --gogoslick_out=. liquidStaking.proto
+package systemSmartContracts
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/core/atomic"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/hashing"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/vm"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+const tokenIDKey = "tokenID"
+const noncePrefix = "n"
+
+type liquidStaking struct {
+	eei                      vm.SystemEI
+	sigVerifier              vm.MessageSignVerifier
+	delegationMgrSCAddress   []byte
+	liquidStakingSCAddress   []byte
+	endOfEpochAddr           []byte
+	gasCost                  vm.GasCost
+	marshalizer              marshal.Marshalizer
+	hasher                   hashing.Hasher
+	mutExecution             sync.RWMutex
+	liquidStakingEnableEpoch uint32
+	flagLiquidStaking        atomic.Flag
+}
+
+// ArgsNewLiquidStaking defines the arguments to create the liquid staking smart contract
+type ArgsNewLiquidStaking struct {
+	EpochConfig            config.EpochConfig
+	Eei                    vm.SystemEI
+	DelegationMgrSCAddress []byte
+	LiquidStakingSCAddress []byte
+	EndOfEpochAddress      []byte
+	GasCost                vm.GasCost
+	Marshalizer            marshal.Marshalizer
+	Hasher                 hashing.Hasher
+	EpochNotifier          vm.EpochNotifier
+}
+
+// NewLiquidStakingSystemSC creates a new liquid staking system SC
+func NewLiquidStakingSystemSC(args ArgsNewLiquidStaking) (*liquidStaking, error) {
+	if check.IfNil(args.Eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+	if len(args.DelegationMgrSCAddress) < 1 {
+		return nil, fmt.Errorf("%w for delegation manager sc address", vm.ErrInvalidAddress)
+	}
+	if len(args.EndOfEpochAddress) < 1 {
+		return nil, fmt.Errorf("%w for end of epoch address", vm.ErrInvalidAddress)
+	}
+	if len(args.LiquidStakingSCAddress) < 1 {
+		return nil, fmt.Errorf("%w for liquid staking sc address", vm.ErrInvalidAddress)
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, vm.ErrNilMarshalizer
+	}
+	if check.IfNil(args.Hasher) {
+		return nil, vm.ErrNilHasher
+	}
+	if check.IfNil(args.EpochNotifier) {
+		return nil, vm.ErrNilEpochNotifier
+	}
+
+	l := &liquidStaking{
+		eei:                      args.Eei,
+		delegationMgrSCAddress:   args.DelegationMgrSCAddress,
+		endOfEpochAddr:           args.EndOfEpochAddress,
+		liquidStakingSCAddress:   args.LiquidStakingSCAddress,
+		gasCost:                  args.GasCost,
+		marshalizer:              args.Marshalizer,
+		hasher:                   args.Hasher,
+		liquidStakingEnableEpoch: args.EpochConfig.EnableEpochs.BuiltInFunctionOnMetaEnableEpoch,
```

Comment:
BuiltInFunctionOnMetaEnableEpoch actually represents the enable epoch for liquid staking?

Reply:
yes. There is no reason to enable esdt on metachain without liquid staking

Reply:
ok